### ISSUE TITLE
tpcc loadgen produces foreign key violation

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#18901]
@jordanlewis this happened to me once. Does it look like a server problem or a client problem?

```
[andrei:~/work/src/github.com/cockroachdb/loadgen/tpcc] master 49s 1 ± ./tpcc -concurrency=128
_time______opName__ops/s(inst)__ops/s(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   1s        tpmC         30.0        29.8    637.5    973.1   1006.6   1006.6
   2s        tpmC         34.0        31.9   1208.0   1744.8   1879.0   1879.0
2017/09/28 16:46:22 error in tpmC: pq: foreign key violation: value [0 6 3162] not found in order@primary [o_w_id o_d_id o_id]
```

**Comment by jordanlewis** [comment]
Probably a client issue - weird though I haven't seen this before.

**Comment by vivekmenezes** [comment]
I see this all the time. I'd suggest waiting for andrei's fix to the deadline exceeded issue before tackling this one.

**Comment by jordanlewis** [comment]
Do you guys still see this?

**Comment by vivekmenezes** [comment]
feel free to close if you can't see it

**Comment by nvanbenschoten** [comment]
I do still see this, but it's very rare.

**Comment by jordanlewis** [comment]
This reoccurs often in testing now, especially on larger datasets.

It's not a client error, because the foreign key error happen in the same transaction that creates the referenced row.

A simplified version of what's going on is:

```
BEGIN TRANSACTION;
INSERT INTO "order" (o_w_id, o_d_id, o_id, ...) VALUES (1, 2, 3, ...)
INSERT INTO order_line (ol_w_id, ol_d_id, ol_o_id, ...) VALUES (1, 2, 3, ...)
2017/09/28 16:46:22 error in tpmC: pq: foreign key violation: value [1 2 3] not found in order@primary [o_w_id o_d_id o_id]
```

The current hypothesis, from @bdarnell and @nvanbenschoten, is that:

> It’s possible that a transaction picks its `d_next_o_id` and lies down an intent on the key. It then inserts into `order` with that id. A second transaction comes along and manages to abort the first transaction, so it “steals” the `d_next_o_id` and removes the intent of the first transaction in the `order` table. The first transaction doesn’t realize it has been aborted and looks back at the `order` table during the FK check to find that its previously written key is no longer there. The problem is that when a transaction gets these seemingly consistency violation errors, we never check to see if it has been aborted.

I'm trying to reproduce this hypothesis in a simpler unit test now.

**Comment by bdarnell** [comment]
OK, I see what's going on here and why the AbortSpan isn't helping us. The key is that the ID being allocated from the `district` table is only part of the order's composite key.

The second transaction encounters our intent on the `district` table. It pushes the first transaction, aborts it (this implies that either there's a deadlock or the system is overloaded enough that the first transaction misses a heartbeat), then resolves away the intent on the district table. This sets the abort span on the district table's range, but the first transaction never revisits the district table to encounter it. Because the order ID is only part of our composite PK, there is no conflict between the two transactions' `order` records, so they don't set the abort span on the `order` range. (the two orders may not even be on the same range) 

The only solution I see to this is to re-query our transaction before returning any error, so we can replace it with a TxnAbortedError. This isn't specific to FK checks, We could narrow the race by re-reading the district table before inserting into order_line, but that would slow down the common case and wouldn't completely solve the problem. 

I'm becoming increasingly convinced that trying to classify errors as retryable or not is a fool's errand but if we didn't have this we'd push a large burden onto the application to decide whether or not an FK error is a result of a txn quirk like this or if it means that the txn cannot succeed.

**Comment by bdarnell** [comment]
> I'm becoming increasingly convinced that trying to classify errors as retryable or not is a fool's errand but if we didn't have this we'd push a large burden onto the application to decide whether or not an FK error is a result of a txn quirk like this or if it means that the txn cannot succeed.

As a quick TPCC-specific fix, since we insert an order and its order_lines in the same transaction, we know that any FK violation we see on the order_line insert must be do to this issue, so to unblock the TPCC work we can just transform that into a retryable error.

**Comment by spencerkimball** [comment]
Something is missing from this explanation. Why wouldn't the foreign key we expect to be there not still be there? If something blew away our inserted foreign key, we'd expect the abort span to be there on the check of the foreign key. I assume I just don't understand the problem, but my belief in this scenario is that the insertion into order succeeds, which means the check from the insertion into order line should see either 1) the inserted `order` entry, or 2) an abort span caused by the forced intent resolve of the inserted `order` entry.

**Comment by bdarnell** [comment]
D'oh, you're right. We'd either see the inserted record or the abort span entry; we can't get a successful result where the record is missing.

**Comment by spencerkimball** [comment]
Are the two inserts for `order` and `order_line` sent as part of a single batch? How do we check foreign key constraints exactly? Could someone point me at the code?

**Comment by bdarnell** [comment]
Separate batches ([starting here](https://github.com/cockroachdb/cockroach/blob/d6980fe993b9646ac7479b1472b84a7270fc8321/pkg/workload/tpcc/new_order.go#L161)). FK checks are [here](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/check.go#L76)

**Comment by jordanlewis** [comment]
I believe that latter link is actually how `SCRUB` validates an entire index. The pathway that gets run during insert is here: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/fk.go#L325

**Comment by spencerkimball** [comment]
Not for nothing but those two requests should be one batch. You could also imagine we might one day suspend the foreign key check if the referee is inserted in the same batch.

**Comment by spencerkimball** [comment]
Oh scratch that...I wasn't thinking clearly. To avoid the check we'd need to do some fancy footwork at the executor level.

**Comment by jordanlewis** [comment]
There was some similar discussion about what we can avoid doing for FK checks in #15157.

**Comment by spencerkimball** [comment]
The way we do foreign key checking is going to prevent batching of inserts from being performant. We check the foreign key referees synchronously on every inserted row. Ouch.

**Comment by jordanlewis** [comment]
Yeah, agreed that that behavior is horrible. We're going to fix it in 2.1. But batching them wouldn't fix whatever is going on here, right?

**Comment by spencerkimball** [comment]
Nope. Just making a comment.
